File: src/main/java/catwalks/part/PartScaffold.java
package catwalks.part;

import catwalks.Const;
import catwalks.EnumCatwalkMaterial;
import catwalks.block.EnumCatwalkMaterialOld;
import catwalks.register.ItemRegister;
import catwalks.util.meta.ArrayProp;
import catwalks.util.meta.IDirtyable;
import catwalks.util.meta.MetaStorage;
import mcmultipart.MCMultiPartMod;
import mcmultipart.multipart.IMultipart;
import mcmultipart.multipart.ISolidPart;
import mcmultipart.multipart.Multipart;
import mcmultipart.raytrace.PartMOP;
import net.minecraft.block.state.BlockStateContainer;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.network.PacketBuffer;
import net.minecraft.util.BlockRenderLayer;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.math.AxisAlignedBB;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by TheCodeWarrior
 */
public class PartScaffold extends Multipart implements ISolidPart, IDirtyable {
	public static final String ID = Const.MODID + ":scaffold";
	
	protected List<AxisAlignedBB> selectionBoxes = new ArrayList<>();
	
	protected static MetaStorage.Allocator allocator = new MetaStorage.Allocator();
	public static ArrayProp<EnumCatwalkMaterial> MATERIAL = allocator.allocateArray("material", EnumCatwalkMaterial.values(), 7);
	
	protected MetaStorage storage = new MetaStorage(allocator, this);
	
	public PartScaffold() {
		double p = 1/16f, P = 1-p;
		double t = 4/16f, T = 1-t;
		selectionBoxes.add(new AxisAlignedBB(0, 0, 0, 1, 1, 1));
	}
	
	public void setCatwalkMaterial(EnumCatwalkMaterial value) {
		if(value != getCatwalkMaterial())
			markDirty();
		MATERIAL.set(storage, value);
	}
	public EnumCatwalkMaterial getCatwalkMaterial() {
		return MATERIAL.get(storage);
	}
	
	@Override
	public NBTTagCompound writeToNBT(NBTTagCompound tag) {
		NBTTagCompound sup = super.writeToNBT(tag);
		sup.setByteArray("m", storage.toByteArray());
		return sup;
	}
	
	@Override
	public void readFromNBT(NBTTagCompound tag) {
		super.readFromNBT(tag);
		storage.fromByteArray(tag.getByteArray("m"));
	}
	
	@Override
	public void writeUpdatePacket(PacketBuffer buf) {
		super.writeUpdatePacket(buf);
		buf.writeByteArray(storage.toByteArray());
	}
	
	@Override
	public void readUpdatePacket(PacketBuffer buf) {
		super.readUpdatePacket(buf);
		storage.fromByteArray(buf.readByteArray());
	}
	
	@Override
	public boolean canRenderInLayer(BlockRenderLayer layer) {
		return layer == getCatwalkMaterial().LAYER;
	}
	
	//region Here be blockstates
	
	@Override
	public BlockStateContainer createBlockState() {
		return new BlockStateContainer(MCMultiPartMod.multipart, Const.MATERIAL_META);
	}
	
	@Override
	public IBlockState getActualState(IBlockState state) {
		return state.withProperty(Const.MATERIAL_META, MATERIAL.get(storage));
	}
	
	//endregion
	
	@Override
	public boolean isSideSolid(EnumFacing side) {
		return true;
	}
	
	@Override
	public void addSelectionBoxes(List<AxisAlignedBB> list) {
		list.addAll(selectionBoxes);
	}
	
	@Override
	public void addCollisionBoxes(AxisAlignedBB mask, List<AxisAlignedBB> list, Entity collidingEntity) {
		AxisAlignedBB axisalignedbb = new AxisAlignedBB(0, 0, 0, 1, 1, 1);
		
		if (mask.intersectsWith(axisalignedbb))
		{
			list.add(axisalignedbb);
		}
	}
	
	@Override
	public boolean occlusionTest(IMultipart part) {
		return super.occlusionTest(part) || part instanceof PartScaffold;
	}
	
	@Override
	public void markDirty() {
		super.markDirty();
	}
	
	@Override
	public ItemStack getPickBlock(EntityPlayer player, PartMOP hit) {
		return new ItemStack(ItemRegister.scaffold, 1, getCatwalkMaterial().ordinal());
	}
}

File: src/main/java/catwalks/part/PartScaffold.kt
package catwalks.part

import catwalks.Const
import catwalks.EnumCatwalkMaterial
import catwalks.block.BlockScaffolding
import catwalks.part.data.ScaffoldRenderData
import catwalks.register.ItemRegister
import catwalks.util.GeneralUtil
import catwalks.util.meta.IDirtyable
import catwalks.util.meta.MetaStorage
import mcmultipart.MCMultiPartMod
import mcmultipart.multipart.IMultipart
import mcmultipart.multipart.ISolidPart
import mcmultipart.multipart.Multipart
import mcmultipart.raytrace.PartMOP
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.PacketBuffer
import net.minecraft.util.BlockRenderLayer
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.AxisAlignedBB
import net.minecraftforge.common.property.ExtendedBlockState
import net.minecraftforge.common.property.IExtendedBlockState
import java.util.*

/**
 * Created by TheCodeWarrior
 */
class PartScaffold : Multipart(), ISolidPart, IDirtyable {

    protected var selectionBoxes: MutableList<AxisAlignedBB> = ArrayList()

    protected var storage = MetaStorage(allocator, this)

    init {
        val p = (1 / 16f).toDouble()
        val P = 1 - p
        val t = (4 / 16f).toDouble()
        val T = 1 - t
        selectionBoxes.add(AxisAlignedBB(0.0, 0.0, 0.0, 1.0, 1.0, 1.0))
    }

    var catwalkMaterial: EnumCatwalkMaterial
        get() = MATERIAL.get(storage)
        set(value) {
            if (value != catwalkMaterial)
                markDirty()
            MATERIAL.set(storage, value)
        }

    override fun writeToNBT(tag: NBTTagCompound): NBTTagCompound {
        val sup = super.writeToNBT(tag)
        sup.setByteArray("m", storage.toByteArray())
        return sup
    }

    override fun readFromNBT(tag: NBTTagCompound?) {
        super.readFromNBT(tag)
        storage.fromByteArray(tag!!.getByteArray("m"))
    }

    override fun writeUpdatePacket(buf: PacketBuffer?) {
        super.writeUpdatePacket(buf)
        buf!!.writeByteArray(storage.toByteArray())
    }

    override fun readUpdatePacket(buf: PacketBuffer?) {
        super.readUpdatePacket(buf)
        storage.fromByteArray(buf!!.readByteArray())
    }

    override fun canRenderInLayer(layer: BlockRenderLayer?): Boolean {
        return layer == catwalkMaterial.LAYER
    }

    //region Here be blockstates

    override fun createBlockState(): BlockStateContainer {
        return ExtendedBlockState(MCMultiPartMod.multipart, arrayOf(Const.MATERIAL), arrayOf(Const.SCAFFOLD_RENDER_DATA))
    }

    override fun getActualState(state: IBlockState): IBlockState {
        return state.withProperty(Const.MATERIAL, MATERIAL.get(storage))
    }

    override fun getExtendedState(state: IBlockState): IBlockState {
        if(state !is IExtendedBlockState)
            return state

        val arr = BooleanArray(EnumFacing.values().size) { true }

        for(dir in EnumFacing.values()) {
            val poff = pos.offset(dir)
            if(world.isSideSolid(poff, dir.opposite)) {
                arr[dir.ordinal] = false
            }
            if(arr[dir.ordinal]) {
                val adjacent = world.getBlockState(poff)
                val block = adjacent.block
                if(block is BlockScaffolding) {
                    if(adjacent.getValue(Const.MATERIAL) == catwalkMaterial) {
                        arr[dir.ordinal] = false
                    }
                }
            }
            if(arr[dir.ordinal]) {
                val part = GeneralUtil.getPart(PartScaffold::class.java, world, poff)
                if(part != null && part.catwalkMaterial == catwalkMaterial) {
                    arr[dir.ordinal] = false
                }
            }
        }

        return state.withProperty(Const.SCAFFOLD_RENDER_DATA, ScaffoldRenderData(arr))
    }

    //endregion

    override fun isSideSolid(side: EnumFacing): Boolean {
        return true
    }

    override fun addSelectionBoxes(list: MutableList<AxisAlignedBB>?) {
        list!!.addAll(selectionBoxes)
    }

    override fun addCollisionBoxes(mask: AxisAlignedBB?, list: MutableList<AxisAlignedBB>?, collidingEntity: Entity?) {
        val axisalignedbb = AxisAlignedBB(0.0, 0.0, 0.0, 1.0, 1.0, 1.0)

        if (mask!!.intersectsWith(axisalignedbb)) {
            list!!.add(axisalignedbb)
        }
    }

    override fun occlusionTest(part: IMultipart?): Boolean {
        return super.occlusionTest(part) || part is PartScaffold
    }

    override fun markDirty() {
        super.markDirty()
    }

    override fun getPickBlock(player: EntityPlayer?, hit: PartMOP?): ItemStack {
        return ItemStack(ItemRegister.scaffold, 1, catwalkMaterial.ordinal)
    }

    companion object {
        val ID = Const.MODID + ":scaffold"

        protected var allocator = MetaStorage.Allocator()
        var MATERIAL = allocator.allocateArray("material", EnumCatwalkMaterial.values(), 7)
    }
}
